### PR TITLE
DAOS-11652 control: Implement systemd Ready notification for daos_agent.

### PR DESCRIPTION
**Pull request by ravalsam**
daos_agent will send the ready notification after unix sockets is up, which will allow user to mount the DFS container via systemd service dependent on daos_agent.

 - Tested with Fuse automount where mount service is depend on daos_agent. User can successfuly mount the fuse which will auto start the daos_agent first.
 - Tested systemd stop daos_agent service which will unmount the fuse as fuse is depend on daos_agent.

Signed-off-by: Samir Raval <samir.raval@intel.com>

### CODE REVIEW

**Review by mjmac — CHANGES_REQUESTED** [review]
Looks good, nice work. I have one inline comment to address, and also a request that you rename `util.go` -> `notify.go`. This new `systemd` package will likely be used for other things, so it's good to keep the files well-named.

One other thing is that we like to add unit test coverage for new code. Please consider the following simple unit test (could be named notify_test.go):

```go
//
// (C) Copyright 2022 Intel Corporation.
//
// SPDX-License-Identifier: BSD-2-Clause-Patent
//

package systemd_test

import (
	"errors"
	"net"
	"os"
	"path/filepath"
	"testing"

	"github.com/daos-stack/daos/src/control/common/test"
	"github.com/daos-stack/daos/src/control/lib/systemd"
)

func Test_Systemd_SdNotify(t *testing.T) {
	testStr := "TEST=1"
	setupSock := func(t *testing.T) string {
		t.Helper()
		sockFile := filepath.Join(t.TempDir(), "notify.sock")
		l, err := net.Listen("unixgram", sockFile)
		if err != nil {
			t.Fatal(err)
		}
		t.Cleanup(func() {
			l.Close()
		})
		return sockFile
	}

	for name, tc := range map[string]struct {
		sockFile string
		expErr   error
	}{
		"success": {
			sockFile: setupSock(t),
		},
		"no socket": {
			expErr: systemd.ErrSdNotifyNoSocket,
		},
		"bad socket": {
			sockFile: "/dev/null",
			expErr: &net.OpError{
				Err: errors.New("connect: connection refused"),
			},
		},
	} {
		t.Run(name, func(t *testing.T) {
			os.Setenv("NOTIFY_SOCKET", tc.sockFile)
			defer os.Unsetenv("NOTIFY_SOCKET")

			err := systemd.SdNotify(testStr)
			test.CmpErr(t, tc.expErr, err)
		})
	}
}
```

**Review by mjmac — APPROVED** [review]

File: src/control/cmd/daos_agent/start.go [hunk 1]
```diff
@@ -117,6 +120,10 @@ func (cmd *startCmd) Execute(_ []string) error {
 
 	cmd.Debugf("startup complete in %s", time.Since(startedAt))
 	cmd.Infof("%s (pid %d) listening on %s", versionString(), os.Getpid(), sockPath)
+	if err := systemd.Ready(); err != nil && err != systemd.ErrSdNotifyNoSocket {
```
Comment (tanabarr) [inline src/control/cmd/daos_agent/start.go:123]: if no socket exists don't we want to notify or at least debug log that the notify wasn't sent?

**Review by tanabarr — APPROVED** [review]
nice work @ravalsam, didn't know you wrote golang!

File: src/control/cmd/daos_agent/start.go (hunk 1 above)
Comment (mjmac) [reply src/control/cmd/daos_agent/start.go:123]: We want to support the non-systemd use case. It's true that there could be a bug where the agent was launched by systemd but there was no socket, but this seems like a pretty rare situation and not worth the log spam in the non-systemd use case.

**Pull request merged**